### PR TITLE
[WIP] Coreset for Kmeans and GaussianMixture clustering

### PR DESCRIPTION
**Pull request by remiadon**
Introduce a Coreset Meta Estimator, that samples a subset of the original data (keeping the original shape) and pass this sample to a scikit-learn estimator

TODO
- [ ] unit tests
- [ ] documentation
- [ ] .fit() returns the scikit-learn estimator, make it returning `self` (the dask meta estimator)
- [ ] validate with plot and comparisons with other clustering algorithms, as in [this page from scikit-learn](https://scikit-learn.org/stable/auto_examples/cluster/plot_cluster_comparison.html#sphx-glr-auto-examples-cluster-plot-cluster-comparison-py))
- [ ] compare Coreset + sklearn.KMeans with dask_ml.KMeans

### CODE REVIEW

**Comment by remiadon** [comment]
Currently I made this comparison, taking [this sklearn doc](https://scikit-learn.org/stable/auto_examples/cluster/plot_cluster_comparison.html#sphx-glr-auto-examples-cluster-plot-cluster-comparison-py) as a baseline


Unknown-3

The original data has 150k rows and 2 features, 200 points are sampled with the Coreset class

*Notes*
I experienced very poor running time performances with small chunks=(100, 2) when building dask.array 
The current graph show running time with a chunk size of (10_000, 2)

**Comment by remiadon** [comment]
Here is a global view of the learning process

dask-coreset

**Comment by TomAugspurger** [comment]
Thanks for continuing to work on this.

> I experienced very poor running time performances with small chunks=(100, 2) when building dask.array

Depending on the scheduler there's a roughly 10-200 microsecond per task overhead (https://docs.dask.org/en/latest/institutional-faq.html?highlight=overhead#how-well-does-dask-scale-what-are-dask-s-limitations).

Can you think of anyone from the scikit-learn side who would be able to review this? I'm not too familiar with the coreset idea / Gaussian mixing. When you're ready, I can take a look at things from the Dask side.

**Comment by remiadon** [comment]
@TomAugspurger my pleasure. Concerning the review from the sklearn side, I think that Jérémie de Boisberranger worked on [a set of runtime/memory benchmarks for KMeans](https://github.com/jeremiedbb/K-means), whose results have been presented at the 2019 scikit-learn consortium.

Concerning interactions with the sklearn.GaussianMixture, I guess we can ask @gmaze, who initially raised the issue on adding Gaussian Mixture Models. He certainly is more informed than I am on how we should validate that our Coreset class can be used (with care ?) with GaussianMixture. My only intuition is that the sklearn.GaussianMixture uses KMeans internally to init the cluster centers (see the `init_params` parameter). The plot at the top of this page visually confirms that the `Coreset(GaussianMixture)` instance succeeds in clustering the flattened forms in the middle, where KMeans fail. 

Also, I would need help on how to properly compare runtime with sklearn. My current benchmark takes the sklearn.KMeans class as an opponent, but sklearn has a [MiniBatchKMeans](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.MiniBatchKMeans.html) which is meant to be faster.
This is to be sure that
 - `Coreset(Kmeans()).fit()` runs as fast as `MiniBatchKmeans` for datasets that hold in memory
 - `Coreset(Kmeans()).fit()` can scale to out-of-memory datasets and still deliver good quality clustering
